pkg/liquidity-mining/test/L2BalancerPseudoMinter.test.ts

```diff
@@ -0,0 +1,325 @@
+        await expectEvent.inIndirectReceipt(receipt, gauge.interface, 'UserCheckpoint', {
+          user: user.address,
+        });
+      });
+    });
+  });
+
+  describe('mintMany', () => {
+    const mockCheckpointStep = fp(1);
+    let gauges: Contract[];
+    let gaugeAddresses: string[];
+
+    sharedBeforeEach('setup factory and fund pseudo minter', async () => {
+      await vault.grantPermissionsGlobally([await actionId(pseudoMinter, 'addGaugeFactory')], admin);
+      await pseudoMinter.connect(admin).addGaugeFactory(gaugeFactory.address);
+      await BAL.connect(admin).mint(pseudoMinter.address, fp(100));
+
+      gauges = await Promise.all(
+        range(3).map(() => {
+          return deployGauge(gaugeFactory);
+        })
+      );
+      gaugeAddresses = gauges.map((gauge) => gauge.address);
+    });
+
+    it('reverts with valid gauge from invalid factory', async () => {
+      await gauges[0].setMockFactory(ANY_ADDRESS);
+      await expect(pseudoMinter.connect(user).mintMany(gaugeAddresses)).to.be.revertedWith('INVALID_GAUGE_FACTORY');
+    });
+
+    it('reverts with malicious gauge prentending to come from a valid factory', async () => {
+      const maliciousGauge = await deploy('MockChildChainGauge', { args: ['test'] });
+      await maliciousGauge.setMockFactory(gaugeFactory.address);
+      const gaugesWithMalicious = [...gaugeAddresses, maliciousGauge.address];
+      await expect(pseudoMinter.connect(user).mintMany(gaugesWithMalicious)).to.be.revertedWith('INVALID_GAUGE');
+    });
+
+    context('when the amount of tokens to transfer is greater than 0', () => {
+      let receipt: ContractReceipt;
+
+      sharedBeforeEach('ensure there will be tokens to transfer after next checkpoint', async () => {
+        // Accounting on the gauge will increase by this amount on every checkpoint.
+        await Promise.all(gauges.map((gauge) => gauge.setMockCheckpointStep(mockCheckpointStep)));
+        receipt = await (await pseudoMinter.connect(user).mintMany(gaugeAddresses)).wait();
+      });
+
+      it('transfers the right amount of tokens to the user in a single transfer', async () => {
+        await expectTransferEvent(
+          receipt,
+          { from: pseudoMinter.address, to: user.address, value: mockCheckpointStep.mul(gauges.length) },
+          BAL
+        );
+      });
+
+      it('updates total minted amount for the user for each gauge', async () => {
+        const totalsBefore = await Promise.all(gauges.map((gauge) => pseudoMinter.minted(user.address, gauge.address)));
+        await pseudoMinter.connect(user).mintMany(gaugeAddresses);
+        const totalsAfter = await Promise.all(gauges.map((gauge) => pseudoMinter.minted(user.address, gauge.address)));
+        expect(totalsAfter).to.be.deep.eq(totalsBefore.map((totalBefore) => totalBefore.add(mockCheckpointStep)));
+      });
+
+      it('emits an event per gauge', async () => {
+        await Promise.all(
```

Comment:
`inReceipt` is not an async function, so you don't need to do this.

Reply:
:sweat_smile: I got carried out.